src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts

```diff
@@ -0,0 +1,747 @@
+/* eslint-disable prettier/prettier */
+import { assert, unreachable } from '../../../../../../common/util/util.js';
+import { EncodableTextureFormat } from '../../../../../format_info.js';
+import { float32ToUint32 } from '../../../../../util/conversion.js';
+import { align, clamp, hashU32, lerp, quantizeToF32 } from '../../../../../util/math.js';
+import {
+  kTexelRepresentationInfo,
+  PerTexelComponent,
+  TexelRepresentationInfo,
+} from '../../../../../util/texture/texel_data.js';
+import { TexelView } from '../../../../../util/texture/texel_view.js';
+import { createTextureFromTexelView } from '../../../../../util/texture.js';
+import { reifyExtent3D } from '../../../../../util/unions.js';
+
+export function createRandomTexelView(info: {
+  format: GPUTextureFormat;
+  size: GPUExtent3D;
+}): TexelView {
+  const rep = kTexelRepresentationInfo[info.format as EncodableTextureFormat];
+  const generator = (coords: Required<GPUOrigin3DDict>): Readonly<PerTexelComponent<number>> => {
+    const texel: PerTexelComponent<number> = {};
+    for (const component of rep.componentOrder) {
+      const rnd = hashU32(coords.x, coords.y, coords.z, component.charCodeAt(0));
+      const normalized = clamp(rnd / 0xffffffff, { min: 0, max: 1 });
+      texel[component] = lerp(rep.numericRange!.finiteMin, rep.numericRange!.finiteMax, normalized);
+    }
+    return quantize(texel, rep);
+  };
+  return TexelView.fromTexelsAsColors(info.format as EncodableTextureFormat, generator);
+}
+
+export type vec2 = [number, number];
+export type vec3 = [number, number, number];
+export type vec4 = [number, number, number, number];
+export type Dimensionality = number | vec2 | vec3;
+
+type TextureCallArgKeys = keyof TextureCallArgs<number>;
+const kTextureCallArgNames: TextureCallArgKeys[] = [
+  'coords',
+  'mipLevel',
+  'arrayIndex',
+  'ddx',
+  'ddy',
+  'offset',
+];
+
+export interface TextureCallArgs<T extends Dimensionality> {
+  coords?: T;
+  mipLevel?: number;
+  arrayIndex?: number;
+  ddx?: T;
+  ddy?: T;
+  offset?: T;
+}
+
+export interface TextureCall<T extends Dimensionality> extends TextureCallArgs<T> {
+  builtin: 'textureSample' | 'textureLoad';
+  coordType: 'f';
+}
+
+function toArray(coords: Dimensionality): number[] {
```

Comment:
nit: possibly just require callers to use `[number]` instead of `number` for scalars, simplifying this?

Reply:
I don't think it's that simple to change it. It used for all args, many of which are more natural to think of as `number` instead of `[number]`. I spent a couple of hours trying to get it work and failed. Maybe we can do this later if it's important?

Reply:
It's not important, just thought it would remove some of the special casing.